snuba/clickhouse/formatter.py

```diff
@@ -0,0 +1,95 @@
+from typing import Optional, Sequence
+from snuba.query.expressions import (
+    Column,
+    CurriedFunctionCall,
+    Expression,
+    ExpressionVisitor,
+    FunctionCall,
+    Literal,
+)
+from snuba.query.parsing import ParsingContext
+from snuba.util import escape_alias, escape_col, escape_string
+
+# Tokens used when formatting. Defining them as costant
+# will make it easy (if/when needed) to make some changes
+# to the layout of the output.
+NULL = "NULL"
+COMMA = ","
+SPACE = " "
+TRUE = "true"
+FALSE = "false"
+QUOTE = "'"
+OPEN_BRACKET = "("
+CLOSED_BRACKET = ")"
+DOT = "."
+
+
+class ClickhouseExpressionFormatter(ExpressionVisitor[str]):
+    """
+    This Visitor implementation is able to format one expression in the Snuba
+    Query for Clickhouse.
+
+    The only state mainained is the PrasingContext, which allows us to resolve
+    aliases and can be reused when formatting multiple expressions.
+
+    When passing an instance of this class to the accept method of
+    the visited expression, the return value is the formatted string.
+
+    This Formatter produces a properly escaped string. The result should never
+    be further escaped. This should be the only place where expression
+    escaping happens as it is done by each method that formats a specific
+    type of expression.
+    """
+
+    def __init__(self, parsing_context: Optional[ParsingContext] = None) -> None:
+        self.__parsing_context = parsing_context or ParsingContext()
```

Comment:
Kind of off-topic and/or out of scope, but we should remember to rename this class (`ParsingContext` when parsing and formatting are completely separate, since I don't think that this class will be used in parsing any longer.